### DISCUSSION TITLE
请问下，如何打一个正常的windows包

### DISCUSSION THREAD

**Discussion by constly** [ejoy/ant#145] [General]
我希望打一个标准windows包，即直接将filepack生成的 bundle文件 + exe文件拷贝给其他人，他们就能正常启动运行。

目前我看如果以-rt启动游戏，会强行去连fileserver。 

实际上我希望最理想的情况是：exe既支持直接读取bundle文件，也支持读取未打包的lua/资源文件。即优先读取包外文件，包外如果没有就去读bundle文件。

不知道怎么样改才能做到这点。

**Comment by cloudwu** [comment]
目前是没有为 windows 设计专门的发布版本的。但可以用 ios 一致的流程，和你说的不太一样。

现在可以把相关的所有文件都打包成 zip ，zip 里是 hash 名的文件。这个 zip 相当于 fileserver 的 cache 的一部分。所以 -rt 启动游戏后，如果发现 zip 包，就会读包里的文件。服务器应该是可选的。

如果按 ios 的发布流程，还可以把新增和改变的文件也打成 zip ，用 http 协议下载，补充到本地 cache 中。通过这种补丁方式下载也用不需要连接 fileserver 。

但是，目前还不支持读非 hash 文件名的外部文件。之后应该会设计一个基于本地文件系统的 mod 让运行模式的 vfs 可以访问。但是，对于需要编译的文件，可能运行模式还不能正确工作。

**Reply by constly** [reply]
我看了输出，发现会去：C:\Users\user_name/Saved Games/ant/internal/00.zip 读取。我把文件放这里确实就可以正常进了

**Reply by constly** [reply]
这个还是不太方便，能否支持读取exe同级目录下的文件呢， 或者exe所在目录固定有个比如叫bundles的文件夹，然后zip就放里面

**Reply by cloudwu** [reply]
这里的路径是想在 windows 下模拟 ios 环境的目录结构。这个接下来会调整。

**Comment by actboy168** [comment]
总的来说需要一个新的io服务。之前有个想法是把现在的两个io服务合并，然后可以配置开启哪些特性，然后可以配置出现有的两种io服务，也可以自己配置其他特性组合的io服务

**Reply by constly** [reply]
看来一时半会还没法。 不过目前我这边 最希望的 还是可以打出 独立的windows包，这样我用ant写的工具就可以给其他人用了。

**Reply by cloudwu** [reply]
如果是支持本地 mod ，目前其实是可以做的。和现在官方编辑器一样的处理方式：

启动后，使用 bee.filesystem 读本地目录，遍历你设定的 mod 目录，把这些文件通过内存文件系统写入 vfs 。例子可以参考：

https://github.com/ejoy/ant/blob/master/test/vfsmem/main.lua

这些 memfs 接口写入 vfs 路径的数据优先级最高。

我觉得现在需要解决的是调整一下 zip 包的路径，因为这是发布文件，放在 `C:\Users\user_name` 下的确不合适。调整到游戏目录更好一些，@actboy168 可以考虑优先改一下这个。

**Reply by cloudwu** [reply]
使用 memfs 而不是改 io 服务来实现 mod ，我觉得更好。启动的时候主动遍历 mod 目录，按需要加载好，这样可以避免各种不必要的未知因素。比如游戏运行过程中 mod 目录内容改变等等。

**Reply by constly** [reply]
这个有意思，我后面尝试下

**Comment by actboy168** [comment]
改了windows/macos上，zip包的位置。

另一个可能有用的功能，这个路径`{exe目录}/external/boot.settings`有个配置文件，可以修改io服务连接文件服务器的行为，建一个空文件的效果就是不连接文件服务器。

**Reply by constly** [reply]
好的，谢谢！ boot.settings 这个很有用，因为我发现即使本地有zip文件，exe也会尝试去连fileserver，要等到socket连接超时，流程才会继续走下去。有了boot.settings启动时就不用卡这几百毫秒了。

**Comment by constly** [comment]
> 如果是支持本地 mod ，目前其实是可以做的。和现在官方编辑器一样的处理方式：
> 启动后，使用 bee.filesystem 读本地目录，遍历你设定的 mod 目录，把这些文件通过内存文件系统写入 vfs 。例子可以参考：
> https://github.com/ejoy/ant/blob/master/test/vfsmem/main.lua
> 这些 memfs 接口写入 vfs 路径的数据优先级最高。
> 


我试了下， test/vfsmem/main.lua 这个测试目前跑不通
```lua
D:\ant\ant>"bin/msvc/release/ant.exe" "test/vfsmem/main.lua" 
[2024-05-12 14:40:45.30][INFO ]( root ) startup 1
[2024-05-12 14:40:45.30][INFO ]( io ) startup 2
[2024-05-12 14:40:45.30][INFO ]( ant.engine|timer ) startup 3
[2024-05-12 14:40:45.33][INFO ]( ant.engine|logger ) startup 4
[2024-05-12 14:40:45.33][INFO ]( /main.lua ) startup 5
[2024-05-12 14:40:45.33][INFO ]( /main.lua ) quit.
[2024-05-12 14:40:45.33][ERROR]( root ) Root fatal:     attempt to call a nil value (field 'reader_new')
stack traceback:
        ( service:2 )
        D:/ant/ant/pkg/ant.vfs/memory.lua:97: in function <D:/ant/ant/pkg/ant.vfs/memory.lua:88>
        (...tail calls...)
        ( service:5 )
        /engine/firmware/ltask_service.lua:222: in upvalue 'rethrow_error'
        D:/ant/ant/test/vfsmem/main.lua:16: in local 'func'
        [C]: in function 'xpcall'
        ( service:5 )
        /engine/firmware/ltask_service.lua:803: in local 's'
        ( service:1 )
        /engine/firmware/ltask_service.lua:222: in upvalue 'rethrow_error'
        engine/firmware/ltask_root.lua:126: in function <engine/firmware/ltask_root.lua:118>
        (...tail calls...)
        engine/firmware/ltask_root.lua:273: in local 'bootstrap'
        engine/firmware/ltask_root.lua:279: in local 'func'
        [C]: in function 'xpcall'
        ( service:1 )
        /engine/firmware/ltask_service.lua:803: in local 's'
```

报在这一行: 
```lua
local c = vfs.read "/testmem/test.txt"
```

**Reply by cloudwu** [reply]
之前迭代 zip 模块的时候，把 `zip.reader_new()` 去掉了，刚才我补了一个同样的函数，放在 fastio 里面。

https://github.com/ejoy/ant/commit/3df918450fcc56e5f03a89cbd0d42963910f5cda

应该已经修好了。

不过奇怪的是， 编辑器应该相关功能早就出问题了？@aimoonchen 是不是相关代码最近都没运行到？我看编辑器还有不少用到 memfs 的地方。我记得刷新 gltf 内部的贴图，在编辑器里立刻显示出来的功能就是依赖 memfs 做的。

**Reply by constly** [reply]
现在可以了，我还有一个问题：对于这个test/vfsmem/main.lua这个测试，我想遍历目录"/testmem"下有哪些文件
```lua
local memfs = import_package "ant.vfs".memory
memfs.init()
memfs.update("/testmem/test.txt", "testmem.txt")
```
该如何写呢，我试了半天没找到门路 = =

**Reply by cloudwu** [reply]
memfs 只是把数据注入了 vfs ，别的地方看起来就是 vfs 里的文件。所以这里，用标准的 vfs api 即可。vfs 的 api 是在引擎启动就被初始化好了，实现在这里：

https://github.com/ejoy/ant/blob/master/engine/firmware/ltask_initservice.lua#L14-L31

遍历用 `vfs.list(path)` 。

ps. VFS 的确缺少文档，需要补一下。

**Reply by cloudwu** [reply]
不好意思，我记错了。vfs 属于比较底层的模块，所以 api 并没有暴露出来。（虽然你还是可以直接使用它们）

推荐使用 `local fs = require "filesystem"` 使用 filesystem ，这是一个封装过的模块。

https://github.com/ejoy/ant/blob/master/engine/filesystem.lua

建议用 `fs.pairs()` 遍历 vfs 里的目录。

**Reply by cloudwu** [reply]
我觉得你可以看看内置的 web 服务器怎么遍历和查看 vfs 里的文件的：

https://github.com/ejoy/ant/blob/master/pkg/ant.webserver/webvfs.lua

这里展示了怎么用 vfs.read 读文件，怎么用 filesystem 遍历虚拟目录，以及用 bee.filesystem 遍历本地目录。

**Reply by constly** [reply]
```lua
local memfs = import_package "ant.vfs".memory
memfs.init()
memfs.update("/testmem/test.txt", "testmem.txt")

local fs = require "filesystem"
for file, file_status in fs.pairs("/testmem") do 
	print("file ", file, file_status)
end
```
我这样试了下，发现不得行，遍历不出来。

**Reply by cloudwu** [reply]
https://github.com/ejoy/ant/commit/fbd3e32e7556958bd9a9435a8be129e41c510d61  之前实现有问题，我重新实现了一遍。有问题再改。

**Reply by cloudwu** [reply]
目前的限制：无法删除原来文件系统已有的文件，只能增加和替换。这个限制其实是可以去掉的。但不想增加不必要的复杂度。

**Reply by constly** [reply]
OK， 我先用看看~

**Comment by constly** [comment]
我还有一个比较纠结的地方。

我写了一个打包的批处理，执行后会将filepack生成的zip和我的exe拷贝到一个指定的发布目录，
我纠结的点在于我必须在目录中放一个专门的批处理来启动程序，即以-rt启动exe：
```
ant_demo.exe -rt
```
一般情况下我们发布的exe就是正式的，用户在得到安装包后，肯定会去点exe启动程序，通过批处理绕一圈实在是不方便。

在不改引擎源码的情况下，我有办法做到不加额外的批处理，直接支持双击exe即可以运行时方式启动游戏吗~ 
我看启动exe时，会自动执行同目录下的main.lua，不知道可否在里面做文章~

**Reply by cloudwu** [reply]
我今天改了一点启动流程。如果 exe 文件不叫 ant / ant.exe ，就默认启动 rt 模式。另外 rt 模式下，不会读 main.lua 。

**Reply by constly** [reply]
我这边有点麻烦，因为我改了c++，我编译的程序貌似必定不能叫ant，原因是ant已经被引擎的make定义了，这是我的编译脚本[make.lua](https://github.com/constly/ant_demo/blob/main/startup/com/make.lua)。所以我现在更新后，我好像没办法以非rt模式运行了。
实际上我现在会编译两个exe，一个是ant.exe，一个是我自己的ant_demo.exe，不知道我怎么改下编译脚本，可以都编译到ant.exe中去。

**Reply by cloudwu** [reply]
你看看 vaststars 怎么编译的，它也有自己的 C++ 代码，编译出来也是 ant.exe 。不过我觉得可以反过来加一个参数控制强制到开发模式下。

**Reply by cloudwu** [reply]
我看了一下，其实条件判断的是文件名以 ant.exe 结尾，而不是判断的全文件名。所以，你还是可以叫 demo_ant.exe 的 :)

**Reply by constly** [reply]
我想起了，当时貌似是因为我这里用了很多imgui的接口，没法把我的代码编译为dll（否则我需要手动导出相关接口），所以后面我改为编译为独立程序了。
不过现在我已经改为demo_ant.exe了
